src/Stackage/Package/Git/Repository.hs

```diff
@@ -0,0 +1,352 @@
+  writeFile (gitLocalPath </> "refs" </> "tags" </> S8.unpack (G.tagBlob newTag)) (show ref)
+  return $ Just ref
+repoCreateTag _ _ _ = return Nothing
+
+
+-- | Clones the repo if it doesn't exists locally yet, otherwise fetches and
+-- creates a local branch.
+ensureRepository
+  :: String -- ^ Git provider (e.g. "github.com")
+  -> String -- ^ Repository account (e.g. "commercialhaskell")
+  -> GitUser -- ^ User information to be used for commits and tags.
+  -> String -- ^ Repository name (e.g. "all-cabal-files")
+  -> String -- ^ Repository branch, make sure there is a remote branch with that
+            -- name. (e.g. "master")
+  -> FilePath -- ^ Location in the file system where repository should be cloned
+     -- to.
+  -> IO GitInfo
+ensureRepository repoHost repoAccount gitUser repoName repoBranchName repoBasePath = do
+  exists <- doesDirectoryExist repoLocalPath
+  if exists
+    then do
+      run repoLocalPath "git" ["remote", "set-url", "origin", repoAddress]
+      run
+        repoLocalPath
+        "git"
+        ["fetch", "origin", repoBranchName <> ":" <> repoBranchName]
+    else run
+           "."
+           "git"
+           [ "clone"
+           , "--bare"
+           , "--depth=1"
+           , repoAddress
+           , repoLocalPath
+           , "--branch"
+           , repoBranchName
+           ]
+  return
+    GitInfo
+    { gitAddress = repoAddress
+    , gitBranchName = repoBranchName
+    , gitUser = gitUser
+    , gitTagName = Nothing
+    , gitLocalPath = repoLocalPath
+    , gitRootTree = Nothing
+    }
+  where
+    repoLocalPath = repoBasePath </> repoName
+    repoAddress =
+      concat ["git@", repoHost, ":", repoAccount, "/", repoName, ".git"]
+
+
+-- | Signs a commit.
+signCommit :: Git -> String -> G.Commit -> IO G.Commit
+signCommit gitRepo key commit = do
+  signature <- signObject gitRepo key (G.ObjCommit commit)
+  -- Workaround for: https://github.com/vincenthz/hit/issues/35
+  -- Otherwise should be:
+  -- let signatureKey = "gpgsig"
+  --     signature = L.toStrict out
+  let signatureKey = L.toStrict $ L.append "gpgsig " $ L8.takeWhile (/='\n') signature
+      signature' = L.toStrict $ L.tail $ L8.dropWhile (/='\n') signature
+  return
+    commit
+    { commitExtras = commitExtras commit ++ [CommitExtra signatureKey signature']
+    }
+
+-- | Signs a tag
+signTag :: Git -> String -> G.Tag -> IO G.Tag
+signTag gitRepo key tag = do
+  signature <- signObject gitRepo key (G.ObjTag tag)
+  return
+    tag
+    { tagS = S8.intercalate "\n" [tagS tag, L.toStrict signature]
+    }
+
+
+-- | Signs an object (really useful only for either a commit or a tag) using a
+-- program specified in git configuration under 'gpg.program'.
+signObject :: Git -> String -> G.Object -> IO L.ByteString
+signObject gitRepo key obj = do
+  gpgProgram <- maybe "gpg" id <$> configGet gitRepo "gpg" "program"
+  let args = ["-bsau", key]
+  let payload = L.tail $ L.dropWhile (/= 0) $ looseMarshall obj
```

Comment:
Yes, I generally like to see no partial functions in code, since even if you know for a fact that the invariants make it total, you're one bug/refactoring away from having a completely unreadable error message coming from who knows where (hopefully stack traces make that less of a problem in the future).

The ideal is to represent at the type level itself that such a situation cannot occur, but sometimes the cost of such type-level solutions is too high for too little benefit.